Teste la reconnaissance de chats

Import needed modules


In [1]:
import numpy as np

import chats_logreg
import chats_tf
import tensorflow as tf


D:\appli\python\anaconda3\envs\ml\lib\site-packages\h5py\__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters


Test de la propagation-avant pour la regression logistique:

1) Test de la regression logistique "à la main"

In [2]:
# Initialise paramters dictionary W
parameters = chats_logreg.initialize_parameters( n_x=3 )

# Test with small input matrix
X = np.array( [ 
                 [ 0.1, 0.01 ],
                 [ 0.2, 0.02 ],
                 [ 0.3, 0.03 ],
            ], dtype=np.float32 )

## number of samples (columns)
m = X.shape[ 1 ]
## number of lines
n_x = X.shape[ 0 ]

## First test with W=0
A = chats_logreg.forward_propagation( X, parameters )
print( "A:", A )

## Some values to W
parameters[ "W" ] = np.array( [ 2, 3, 4 ] )
A = chats_logreg.forward_propagation( X, parameters )
print( "A:", A )

# Get cost
Y = np.array( [ 
                 [ 1 ],
                 [ 0 ],
                 [ 1 ],
            ], dtype=np.float32 )

cost = chats_logreg.compute_cost( A, Y, WEIGHT=1, m = m )
print( "Cost:", cost )

A: [[0.5 0.5]]
A: [0.88079708 0.549834  ]
Cost: 2.1876003405650217


Expected:
<code>
A: [[0.5 0.5]]
A: [0.88079708 0.549834  ]
Cost: 2.1876003405650217
</code>

2) Test de la régression logistrique TensorFlow

In [3]:
# Create TF variables place holders - do it once

nbUnits = [ 1 ]

X_pl, Y_pl, KEEP_PROB_pl = chats_tf.create_placeholders( n_x=n_x, n_y=1 )
# Initialise parameters
parameters_tf = chats_tf.initialize_parameters( nbUnits=nbUnits, n_x=n_x )


Instructions for updating:
Use the retry module or similar alternatives.


In [4]:
# forward-propagation function
Z_tf = chats_tf.forward_propagation( X_pl, parameters=parameters_tf, nbUnits=nbUnits, n_x=n_x, KEEP_PROB = KEEP_PROB_pl )
# Cost
cost_tf = chats_tf.compute_cost( Z_tf, Y_pl, WEIGHT=1, beta=0, parameters=parameters_tf, nbUnits=nbUnits, n_x=n_x )
print( "Cost tensor:", cost_tf )

## Calculate cost
init = tf.global_variables_initializer()

# Start the session to compute the tensorflow graph
with tf.Session() as sess:
    _ , cost = sess.run( [cost_tf], feed_dict={ X_pl: X, Y_pl: Y.T, KEEP_PROB_pl: 1 } )
                    
print( "Cost:", cost )


Cost tensor: Tensor("Mean_1:0", shape=(), dtype=float32)


FailedPreconditionError: Attempting to use uninitialized value W1
	 [[Node: W1/read = Identity[T=DT_FLOAT, _class=["loc:@W1"], _device="/job:localhost/replica:0/task:0/device:CPU:0"](W1)]]

Caused by op 'W1/read', defined at:
  File "D:\appli\python\anaconda3\envs\ml\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "D:\appli\python\anaconda3\envs\ml\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\ipykernel_launcher.py", line 16, in <module>
    app.launch_new_instance()
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\traitlets\config\application.py", line 658, in launch_instance
    app.start()
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\ipykernel\kernelapp.py", line 486, in start
    self.io_loop.start()
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\tornado\platform\asyncio.py", line 127, in start
    self.asyncio_loop.run_forever()
  File "D:\appli\python\anaconda3\envs\ml\lib\asyncio\base_events.py", line 422, in run_forever
    self._run_once()
  File "D:\appli\python\anaconda3\envs\ml\lib\asyncio\base_events.py", line 1432, in _run_once
    handle._run()
  File "D:\appli\python\anaconda3\envs\ml\lib\asyncio\events.py", line 145, in _run
    self._callback(*self._args)
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\tornado\ioloop.py", line 759, in _run_callback
    ret = callback()
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\tornado\stack_context.py", line 276, in null_wrapper
    return fn(*args, **kwargs)
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\zmq\eventloop\zmqstream.py", line 536, in <lambda>
    self.io_loop.add_callback(lambda : self._handle_events(self.socket, 0))
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\zmq\eventloop\zmqstream.py", line 450, in _handle_events
    self._handle_recv()
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\zmq\eventloop\zmqstream.py", line 480, in _handle_recv
    self._run_callback(callback, msg)
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\zmq\eventloop\zmqstream.py", line 432, in _run_callback
    callback(*args, **kwargs)
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\tornado\stack_context.py", line 276, in null_wrapper
    return fn(*args, **kwargs)
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\ipykernel\kernelbase.py", line 283, in dispatcher
    return self.dispatch_shell(stream, msg)
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\ipykernel\kernelbase.py", line 233, in dispatch_shell
    handler(stream, idents, msg)
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\ipykernel\kernelbase.py", line 399, in execute_request
    user_expressions, allow_stdin)
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\ipykernel\ipkernel.py", line 208, in do_execute
    res = shell.run_cell(code, store_history=store_history, silent=silent)
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\ipykernel\zmqshell.py", line 537, in run_cell
    return super(ZMQInteractiveShell, self).run_cell(*args, **kwargs)
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\IPython\core\interactiveshell.py", line 2662, in run_cell
    raw_cell, store_history, silent, shell_futures)
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\IPython\core\interactiveshell.py", line 2785, in _run_cell
    interactivity=interactivity, compiler=compiler, result=result)
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\IPython\core\interactiveshell.py", line 2903, in run_ast_nodes
    if self.run_code(code, result):
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\IPython\core\interactiveshell.py", line 2963, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-3-34e8c30d54bb>", line 7, in <module>
    parameters_tf = chats_tf.initialize_parameters( nbUnits=nbUnits, n_x=n_x )
  File "C:\Users\frup82455\git\ml\chats\chats\chats_tf.py", line 76, in initialize_parameters
    W_cur = tf.get_variable( "W" + str( i + 1 ), [ nbUnits0[ i + 1 ], nbUnits0[ i ] ], initializer = tf.contrib.layers.xavier_initializer(seed = 1))
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\tensorflow\python\ops\variable_scope.py", line 1297, in get_variable
    constraint=constraint)
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\tensorflow\python\ops\variable_scope.py", line 1093, in get_variable
    constraint=constraint)
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\tensorflow\python\ops\variable_scope.py", line 439, in get_variable
    constraint=constraint)
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\tensorflow\python\ops\variable_scope.py", line 408, in _true_getter
    use_resource=use_resource, constraint=constraint)
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\tensorflow\python\ops\variable_scope.py", line 800, in _get_single_variable
    use_resource=use_resource)
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\tensorflow\python\ops\variable_scope.py", line 2157, in variable
    use_resource=use_resource)
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\tensorflow\python\ops\variable_scope.py", line 2147, in <lambda>
    previous_getter = lambda **kwargs: default_variable_creator(None, **kwargs)
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\tensorflow\python\ops\variable_scope.py", line 2130, in default_variable_creator
    constraint=constraint)
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\tensorflow\python\ops\variables.py", line 235, in __init__
    constraint=constraint)
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\tensorflow\python\ops\variables.py", line 391, in _init_from_args
    self._snapshot = array_ops.identity(self._variable, name="read")
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\tensorflow\python\ops\array_ops.py", line 142, in identity
    return gen_array_ops.identity(input, name=name)
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\tensorflow\python\ops\gen_array_ops.py", line 3052, in identity
    "Identity", input=input, name=name)
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\tensorflow\python\framework\op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\tensorflow\python\framework\ops.py", line 3290, in create_op
    op_def=op_def)
  File "D:\appli\python\anaconda3\envs\ml\lib\site-packages\tensorflow\python\framework\ops.py", line 1654, in __init__
    self._traceback = self._graph._extract_stack()  # pylint: disable=protected-access

FailedPreconditionError (see above for traceback): Attempting to use uninitialized value W1
	 [[Node: W1/read = Identity[T=DT_FLOAT, _class=["loc:@W1"], _device="/job:localhost/replica:0/task:0/device:CPU:0"](W1)]]
